plugins/basic/nexus-repository-docker/src/main/java/org/sonatype/nexus/repository/docker/internal/DockerFacetImpl.java

```diff
@@ -0,0 +1,682 @@
+    tx.setBlob(imagesIndexAsset, imagesIndexAssetName, indexStream, hashAlgorithms, null, payload.getContentType());
+    tx.saveAsset(imagesIndexAsset);
+  }
+
+  @Override
+  @Transactional(retryOn = IllegalStateException.class)
+  public Payload getLayerMetadata(final String layerId) throws IOException {
+    validateLayerId(layerId);
+    StorageTx tx = UnitOfWork.currentTransaction();
+
+    String metadataAssetName = layerMetadataName(layerId);
+
+    Bucket bucket = tx.getBucket();
+    Asset metadataAsset = findAsset(tx, bucket, metadataAssetName);
+    // check that we have the layer and layer content was fully uploaded
+    if (metadataAsset == null || Boolean.TRUE.equals(metadataAsset.formatAttributes().get(P_IN_PROGRESS))) {
+      throw new V1Exception.ImageNotFound();
+    }
+
+    Content content = marshall(metadataAsset, tx.requireBlob(metadataAsset.requireBlobRef()));
+
+    NestedAttributesMap componentAttributes = metadataAsset.formatAttributes();
+    content.getAttributes().set(P_X_DOCKER_SIZE, componentAttributes.get(P_X_DOCKER_SIZE));
+    content.getAttributes().set(P_X_DOCKER_CHECKSUM_PAYLOAD, componentAttributes.get(P_X_DOCKER_CHECKSUM_PAYLOAD));
+
+    return content;
+  }
+
+  @Override
+  public void putLayerMetadata(final String layerId, final Payload content) throws IOException {
+    validateLayerId(layerId);
+    try (TempStreamSupplier streamSupplier = new TempStreamSupplier(content.openInputStream())) {
+      doPutLayerMetadata(layerId, streamSupplier, content);
+    }
+  }
+
+  @Transactional(retryOn = {ONeedRetryException.class, ORecordDuplicatedException.class})
+  protected void doPutLayerMetadata(final String layerId,
+                                    final Supplier<InputStream> streamSupplier,
+                                    final Payload payload)
+      throws IOException
+  {
+    Map<String, Object> data = readLayerMetadata(streamSupplier);
+    if (!data.containsKey("id")) {
+      throw new V1Exception.InvalidLayerMetadata("Missing key 'id' in JSON");
+    }
+    if (!data.get("id").equals(layerId)) {
+      throw new V1Exception.InvalidLayerMetadata("JSON data contains invalid id");
+    }
+    StorageTx tx = UnitOfWork.currentTransaction();
+
+    String metadataAssetName = layerMetadataName(layerId);
+    String parentId = (String) data.get(DockerAttributes.P_PARENT);
+
+    Bucket bucket = tx.getBucket();
+
+    Asset parentAsset = null;
+    List<String> ancestry = null;
+    if (parentId != null) {
+      // check that we have the parent layer and parent layer content was fully uploaded
+      parentAsset = findAsset(tx, bucket, layerMetadataName(parentId));
+      if (parentAsset == null || Boolean.TRUE.equals(parentAsset.formatAttributes().get(P_IN_PROGRESS))) {
+        throw new V1Exception.InvalidLayerMetadata("Image depends on a non existing parent");
+      }
+    }
+    // TODO check that parent asset is accessible via authorisation once we have per assets auth
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)